### PR TITLE
Initialization via Job Packet and Rename DataQueueHandler to DataProvider

### PR DESCRIPTION
**Pull request by Romazes**
#### Description

1. Feature Enhancement: Initialization via Job Packet
  - Implements initialization functionality through **Job packets**, enhancing the efficiency and flexibility of the system.
Provides a streamlined method for initializing processes, improving overall performance and maintainability.

2. Code Refactoring: Renaming DataQueueHandler to DataProvider
  - Renames the **DataQueueHandler** class to **DataProvider** for improved clarity and consistency in naming conventions.
Enhances code readability and understandability, aligning with best practices for naming conventions.

3. Use Globals for credentials access for subscription validation.
  - This follows https://github.com/QuantConnect/Lean/pull/7796

4. Handle history provider returning null
  - This follows https://github.com/QuantConnect/Lean/pull/7804
5. Handle data downloader provider returning null
  - This follows https://github.com/QuantConnect/Lean/pull/7812

#### Related Issue

N/A

#### Motivation and Context

1. Feature Enhancement: Initialization via Job Packet
  - As the system evolves, there is a growing need for a more robust and flexible initialization process. Implementing initialization through Job packets provides an efficient and adaptable solution, enhancing overall system efficiency and maintainability.
2. Code Refactoring: Renaming DataQueueHandler to DataProvider
  - Clear and consistent naming conventions are crucial for improving code maintainability. Renaming DataQueueHandler to DataProvider aligns with best practices, improving code clarity and making it easier for developers to understand and work with the codebase.

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

To ensure the robustness and correctness of the Job initialization process, the PR have been implemented specific test cases.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: QuantConnect.CoinAPI/CoinAPIDataDownloader.cs [hunk 1]
```diff
@@ -64,7 +64,20 @@ public IEnumerable<BaseData> Get(DataDownloaderGetParameters dataDownloaderGetPa
                     dataNormalizationMode: DataNormalizationMode.Raw,
                     tickType: TickType.Trade);
 
-            foreach (var slice in _historyProvider.GetHistory(historyRequests))
+            var history = _historyProvider.GetHistory(historyRequests);
+
+            // historyRequest contains wrong data request
+            if (history == null)
+            {
+                return null;
```
Comment (jhonabreul) [inline QuantConnect.CoinAPI/CoinAPIDataDownloader.cs:59]: I think the downloader could still return empty is the history provider returns null. It's a different use case

File: QuantConnect.CoinAPI/CoinApiDataHistoryProvider.cs [hunk 2]
```diff
@@ -48,39 +48,59 @@ public override IEnumerable<Slice> GetHistory(IEnumerable<HistoryRequest> reques
...
             if (historyRequest.DataType == typeof(QuoteBar))
             {
                 if (!_invalidHistoryDataTypeWarningFired)
                 {
-                    Log.Error("CoinApiDataQueueHandler.GetHistory(): No historical QuoteBars , only TradeBars");
+                    Log.Error("CoinApiDataProvider.GetHistory(): No historical QuoteBars , only TradeBars");
                     _invalidHistoryDataTypeWarningFired = true;
                 }
-                yield break;
+                return null;
```
Comment (jhonabreul) [inline QuantConnect.CoinAPI/CoinApiDataHistoryProvider.cs]: You could keep some flags to log this errors only once and not flood the output. See https://github.com/QuantConnect/Lean.Brokerages.Bitfinex/blob/14f8ba19a3b8fa1fe73da5724b102c877320f488/QuantConnect.BitfinexBrokerage/BitfinexBrokerage.cs#L368-L373

File: QuantConnect.CoinAPI/CoinApiDataProvider.cs [hunk 3]
```diff
@@ -122,6 +123,63 @@ public IEnumerator<BaseData> Subscribe(SubscriptionDataConfig dataConfig, EventH
...
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="apiKey"></param>
+        /// <param name="product"></param>
+        /// <exception cref="ArgumentException"></exception>
+        private void Initialize(string apiKey, string product)
+        {
+            // ValidateSubscription();
```
Comment (jhonabreul) [inline QuantConnect.CoinAPI/CoinApiDataProvider.cs]: Whops

File: QuantConnect.CoinAPI/CoinApiDataProvider.cs [hunk 4]
```diff
@@ -292,14 +353,23 @@ private void SendHelloMessage(IEnumerable<string> subscribeFilter)
...
                 heartbeat = true,
                 subscribe_data_type = _streamingDataType,
                 subscribe_filter_symbol_id = list.ToArray()
             });
 
+            if (!IsConnected && !_client.ConnectedEvent.WaitOne(TimeSpan.FromSeconds(30)))
+            {
+                throw new Exception("Not connected...");
+            }
+            else
```
Comment (jhonabreul) [inline QuantConnect.CoinAPI/CoinApiDataProvider.cs]: The `else` might not be necessary here

**Review by jhonabreul — CHANGES_REQUESTED** [review]
Nice work! Leaving a few comments

File: QuantConnect.CoinAPI/CoinAPIDataDownloader.cs (hunk 1 above)
Comment (jhonabreul) [reply QuantConnect.CoinAPI/CoinAPIDataDownloader.cs:59]: Don't mind this one 👌

Comment (Romazes) [reply QuantConnect.CoinAPI/CoinAPIDataDownloader.cs:59]: https://github.com/QuantConnect/Lean/pull/7812

**Review by jhonabreul — APPROVED** [review]

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**